common/installshield_cab.cpp

```diff
@@ -48,114 +48,177 @@
 #include "common/hash-str.h"
 #include "common/installshield_cab.h"
 #include "common/memstream.h"
+#include "common/substream.h"
+#include "common/ptr.h"
 #include "common/zlib.h"
 
 namespace Common {
 
+namespace {
+
 class InstallShieldCabinet : public Archive {
 public:
-	InstallShieldCabinet(SeekableReadStream *stream, DisposeAfterUse::Flag disposeAfterUse);
-	~InstallShieldCabinet();
+	InstallShieldCabinet();
+
+	bool open(const String &baseName);
+	void close();
 
 	// Archive API implementation
-	bool hasFile(const String &name) const;
-	int listMembers(ArchiveMemberList &list) const;
-	const ArchiveMemberPtr getMember(const String &name) const;
-	SeekableReadStream *createReadStreamForMember(const String &name) const;
+	virtual bool hasFile(const String &name) const;
+	virtual int listMembers(ArchiveMemberList &list) const;
+	virtual const ArchiveMemberPtr getMember(const String &name) const;
+	virtual SeekableReadStream *createReadStreamForMember(const String &name) const;
 
 private:
 	struct FileEntry {
 		uint32 uncompressedSize;
 		uint32 compressedSize;
 		uint32 offset;
 		uint16 flags;
+		uint16 volume;
 	};
 
+	int _version;
 	typedef HashMap<String, FileEntry, IgnoreCase_Hash, IgnoreCase_EqualTo> FileMap;
 	FileMap _map;
-	Common::SeekableReadStream *_stream;
-	DisposeAfterUse::Flag _disposeAfterUse;
-};
+	String _baseName;
 
-InstallShieldCabinet::~InstallShieldCabinet() {
-	_map.clear();
+	String getHeaderName() const;
+	String getVolumeName(uint volume) const;
+};
 
-	if (_disposeAfterUse == DisposeAfterUse::YES)
-		delete _stream;
+InstallShieldCabinet::InstallShieldCabinet() : _version(0) {
 }
 
-InstallShieldCabinet::InstallShieldCabinet(SeekableReadStream *stream, DisposeAfterUse::Flag disposeAfterUse) : _stream(stream), _disposeAfterUse(disposeAfterUse) {
-	// Note that we only support a limited subset of cabinet files
-	// Only single cabinet files and ones without data shared between
-	// cabinets.
+bool InstallShieldCabinet::open(const String &baseName) {
+	// Store the base name so we can generate file names
+	_baseName = baseName;
+
+	// Try to find the file
+	SeekableReadStream *header = SearchMan.createReadStreamForMember(getHeaderName());
+	if (!header)
+		header = SearchMan.createReadStreamForMember(getVolumeName(1));
+
+	if (!header) {
+		close();
+		return false;
+	}
+
+	// Note that we only support v5 and v6 cabinet files
 
 	// Check for the magic uint32
-	uint32 magic = _stream->readUint32LE();
+	uint32 magic = header->readUint32LE();
 	if (magic != 0x28635349) {
 		warning("InstallShieldCabinet::InstallShieldCabinet(): Magic ID doesn't match: expecting %x but got %x", 0x28635349, magic);
-		return;
+		close();
+		return false;
 	}
 
-	uint32 version = _stream->readUint32LE();
-
-	if (version != 0x01000004) {
+	uint32 version = header->readUint32LE();
+
+	switch(version) {
+	case 0x01000004:
```

Comment:
Based on the previous comment, I would suggest to change this switch into something like this:

    int shift = magicBytes[i] >> 24;
    _version = shift == 1 ? (version >> 12) & 0xf : (version & 0xffff) / 100;
    _version = (_version == 0) ? 5 : _version;

The last check for version 0 has been commented out in unshield, but we are treating this version as 5 in our code, so it's needed

Reply:
Done. Based on the unshield code we should now (mostly) support versions 5-13, though I'd need some more test cases to verify if that's actually true.

Reply:
This correctly handles the LSL7 Polish files. (Data1.cab/hdr) They are version 5 with magicBytes 0x01005201. I confirmed in the installer's version strings that it really is version 5.